metadata/src/main/java/org/apache/kafka/metadata/MetadataParser.java

```diff
@@ -0,0 +1,115 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.kafka.metadata;
+
+import org.apache.kafka.common.metadata.MetadataRecordType;
+import org.apache.kafka.common.protocol.ApiMessage;
+import org.apache.kafka.common.protocol.ByteBufferAccessor;
+import org.apache.kafka.common.protocol.ObjectSerializationCache;
+import org.apache.kafka.common.utils.ByteUtils;
+
+import java.nio.ByteBuffer;
+
+public class MetadataParser {
+    public static final int MAX_SERIALIZED_EVENT_SIZE = 32 * 1024 * 1024;
+
+    private static short unsignedIntToShort(int val, String what) {
+        if (val > Short.MAX_VALUE) {
```

Comment:
We should also check `val < 0` here to make sure it is non-negative.

Reply:
It's not necessary here since the value comes from `ByteUtils.readUnsignedVarint`, which won't return a negative.  This is an advantage of using a private function rather than something more general.